Energy_Preserving.jl

```diff
@@ -0,0 +1,521 @@
+        end
+    end
+    return theindex
+end
+
+iswhitespace(c::Char) = isspace(c)
+#this function is the BMinus operator as described in B-Series papers
+function BMinus(array)
+    m = 1
+    for i in array
+        if i == 2
+            m = m + 1
+        end
+    end
+    counter = 0
+    #we save the trees here
+    auxiliar = Array{Int64}(undef, m)
+    l = length(array)
+    auxiliar[m] = l
+    #the number of subtrees
+    n = m-1
+    subtrees = Vector{Any}(undef, n)
+    for i in 2:l
+        if array[i] == 2
+            counter += 1
+            auxiliar[counter] = i
+            if counter == n
+                break
+            end
+        end
+    end
+    for i in 1:n
+        subtrees[i] = array[auxiliar[i]:(auxiliar[i+1]-1)]
+    end
+    push!(subtrees[n],array[l])
+    return subtrees
+end
+         
+
+#generates all equivalent trees
+function equivalent_trees(array)
+    tree = rootedtree(array)
+    l = length(array)
+    #we get the permutations of the array
+    superarray = get_permutations(array)
+    lperm = length(superarray)
+    #eliminate all the arrays except for those for which the first element is equal to 1
+    for i in reverse(1:lperm) 
+        if superarray[i][1] != 1
+            splice!(superarray, i)
+        end
+    end
+    lperm = length(superarray)
+    #now, check if the levelsequence is the level sequence represents a RootedTree:
+    #the condition is that each number m cannot be followed by a number n such that
+    #n-m > 1. If that's the case, eliminate this array with 'splice!'
+    RTree_array = true
+    for i in reverse(1:lperm)
+        for j in 1:(l-1)
+            if (superarray[i][j+1] - superarray[i][j]) > 1
+                RTree_array = false
+            end
+        end
+        if RTree_array == false
+            splice!(superarray, i)
+        end
+        RTree_array = true
+    end
+    lperm = length(superarray)
+    for i in reverse(1:lperm)
+        onetree = rootedtree(superarray[i])
+        isthesame_tree = onetree == tree
+        if isthesame_tree == false
+            splice!(superarray, i)
+        end
+    end
+    repeated_tree = false
+    lperm = length(superarray)
+    #we delete any repeated tree
+    for i in reverse(1:lperm)
+        for j in 1:(i-1)
+            if superarray[i] == superarray[j]
+                repeated_tree = true
+            end
+        end 
+        if repeated_tree == true
+            splice!(superarray, i)
+        end
+        repeated_tree = false
+    end
+
+    return superarray
+end
+
+function get_permutations(arr)
+    perms = collect(permutations(arr))
+    return perms
+end
+
+function num_leafs(a)
+    k = a[end]
+    return k - 1
+end
+
+function add_one_to_left(arrays)
+    l = length(arrays)
+    for i in 1:l
+        arrays[i] = vcat([1], arrays[i])
+    end
+    return arrays
+end
+
+#this function multplies the coefficient for its symmetry factor
+function symfact_normalization(coef,thetrees)
```

Comment:
Let's call this something like `renormalize_bseries`.  Then refactor it to take a BSeries as input and return a dict.